llvm/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp

```diff
@@ -984,7 +1041,7 @@ Value *InstCombinerImpl::SimplifyDemandedUseBits(Value *V, APInt DemandedMask,
 
   // If the client is only demanding bits that we know, return the known
   // constant.
-  if (DemandedMask.isSubsetOf(Known.Zero|Known.One))
+  if (MaySimplifyAsConstant && DemandedMask.isSubsetOf(Known.Zero | Known.One))
```

Comment:
FWIW, it would be legal to return a inttoptr constant for non-zero values. But not sure if it would be a good idea, so staying conservative is reasonable.

Reply:
Ill add that as a todo. Although maybe indicative we want to the flag to be based on the exact intrin as opposed to `isPointer` as different intrins may have different preference. For `@llvm.ptrmask` for example. we can easily represent the most common constant (`null`) by just zeroing the mask so there isn't much benefit to constant return. But for other intrins might be different.

Reply:
Leaving as TODO + made just a pointer check.